File: app/src/main/java/com/coinninja/coinkeeper/cn/dropbit/DropBitCancellationManager.java
package com.coinninja.coinkeeper.cn.dropbit;

import com.coinninja.coinkeeper.cn.wallet.CNWalletManager;
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary;
import com.coinninja.coinkeeper.model.helpers.TransactionHelper;
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient;

import java.util.List;

import javax.inject.Inject;

import retrofit2.Response;

public class DropBitCancellationManager {
    private final SignedCoinKeeperApiClient client;
    private final TransactionHelper transactionHelper;
    private final CNWalletManager cnWalletManager;

    @Inject
    DropBitCancellationManager(SignedCoinKeeperApiClient client, TransactionHelper transactionHelper, CNWalletManager cnWalletManager) {
        this.client = client;
        this.transactionHelper = transactionHelper;
        this.cnWalletManager = cnWalletManager;
    }

    public void markAsCanceled(List<InviteTransactionSummary> summaries) {
        if (summaries == null || summaries.isEmpty()) return;
        for (InviteTransactionSummary invite : summaries) {
            markAsCanceled(invite);
        }
    }

    public void markAsCanceled(InviteTransactionSummary invite) {
        String inviteID = invite.getServerId();

        Response response = client.updateInviteStatusCanceled(inviteID);
        if (response.isSuccessful()) {
            transactionHelper.updateInviteAsCanceled(inviteID);
            cnWalletManager.updateBalances();
        }
    }

    public void markAsCanceled(String id) {
        InviteTransactionSummary inviteTransactionSummary = transactionHelper.getInviteTransactionSummary(id);
        if (inviteTransactionSummary != null)
            markAsCanceled(inviteTransactionSummary);
    }

    public void markUnfulfilledAsCanceled() {
        markAsCanceled(transactionHelper.gatherUnfulfilledInviteTrans());
        cnWalletManager.updateBalances();
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/cn/dropbit/DropBitCancellationManager.kt
package com.coinninja.coinkeeper.cn.dropbit

import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.cn.wallet.CNWalletManager
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary
import com.coinninja.coinkeeper.model.helpers.InviteTransactionSummaryHelper
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient
import javax.inject.Inject

@Mockable
class DropBitCancellationManager @Inject internal constructor(
        internal val client: SignedCoinKeeperApiClient,
        internal val inviteTransactionSummaryHelper: InviteTransactionSummaryHelper,
        internal val cnWalletManager: CNWalletManager
) {

    fun markAsCanceled(summaries: List<InviteTransactionSummary>) {
        for (invite in summaries) {
            markAsCanceled(invite)
        }
    }

    fun markAsCanceled(invite: InviteTransactionSummary) {
        val inviteID = invite.serverId

        val response = client.updateInviteStatusCanceled(inviteID)
        if (response.isSuccessful) {
            inviteTransactionSummaryHelper.cancelInviteByCnId(inviteID)
            cnWalletManager.updateBalances()
        }
    }

    fun markAsCanceled(id: String) {
        inviteTransactionSummaryHelper.getInviteSummaryByCnId(id)?.let {
            markAsCanceled(it)
        }

    }

    fun markUnfulfilledAsCanceled() {
        markAsCanceled(inviteTransactionSummaryHelper.unfulfilledSentInvites)
        cnWalletManager.updateBalances()
    }
}
